sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala

```diff
@@ -0,0 +1,410 @@
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.expressions.{Ascending, Descending}
+import org.apache.spark.sql.catalyst.parser._
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.types.StructType
+
+
+/**
+ * Helper object to parse alter table commands.
+ */
+object AlterTableCommandParser {
+  import ParserUtils._
+
+  /**
+   * Parse the given node assuming it is an alter table command.
+   */
+  def parse(v1: ASTNode): LogicalPlan = {
+    v1.children match {
+      case (tabName @ Token("TOK_TABNAME", _)) :: restNodes =>
+        val tableIdent: TableIdentifier = extractTableIdent(tabName)
+        val partitionSpec = getClauseOption("TOK_PARTSPEC", v1.children)
+        val partition = partitionSpec.flatMap(parsePartitionSpec)
+        matchAlterTableCommands(v1, restNodes, tableIdent, partition)
+      case _ =>
+        throw new AnalysisException(s"Could not parse alter table command: '${v1.text}'")
+    }
+  }
+
+  private def cleanAndUnquoteString(s: String): String = {
+    cleanIdentifier(unquoteString(s))
+  }
+
+  private def parsePartitionSpec(node: ASTNode): Option[Map[String, Option[String]]] = {
+    node match {
+      case Token("TOK_PARTSPEC", partitions) =>
+        val spec = partitions.map {
+          case Token("TOK_PARTVAL", ident :: constant :: Nil) =>
+            (cleanAndUnquoteString(ident.text), Some(cleanAndUnquoteString(constant.text)))
+          case Token("TOK_PARTVAL", ident :: Nil) =>
+            (cleanAndUnquoteString(ident.text), None)
+        }.toMap
+        Some(spec)
+      case _ => None
+    }
+  }
+
+  private def extractTableProps(node: ASTNode): Map[String, Option[String]] = {
+    node match {
+      case Token("TOK_TABLEPROPERTIES", propsList) =>
+        propsList.flatMap {
+          case Token("TOK_TABLEPROPLIST", props) =>
+            props.map {
+              case Token("TOK_TABLEPROPERTY", key :: Token("TOK_NULL", Nil) :: Nil) =>
+                val k = cleanAndUnquoteString(key.text)
+                (k, None)
+              case Token("TOK_TABLEPROPERTY", key :: value :: Nil) =>
+                val k = cleanAndUnquoteString(key.text)
+                val v = cleanAndUnquoteString(value.text)
+                (k, Some(v))
+            }
+        }.toMap
+      case _ =>
+        throw new AnalysisException(
+          s"Expected table properties in alter table command: '${node.text}'")
+    }
+  }
+
+  // TODO: This method is massive. Break it down. Also, add some comments...
+  private def matchAlterTableCommands(
+      node: ASTNode,
+      nodes: Seq[ASTNode],
+      tableIdent: TableIdentifier,
+      partition: Option[Map[String, Option[String]]]): LogicalPlan = {
+    nodes match {
+      case rename @ Token("TOK_ALTERTABLE_RENAME", renameArgs) :: _ =>
```

Comment:
Ah. That is because the node is structured as (take `TOK_ALTERTABLE_RENAME` as an example):

    Token("TOK_TABNAME", _) :: Token("TOK_ALTERTABLE_RENAME", renameArgs) :: Token("TOK_PARTSPEC", _)

As you see, Token("TOK_PARTSPEC", _) is put at the end of the AST. Although I've parsed partspec early, I pass the all nodes after `Token("TOK_TABNAME", _)` to the method `matchAlterTableCommands`. So there is always an additional input that can safely ignore.